File: PreUtil/src/output/java/hard/ShortestPathInAGridWithObstaclesElimination.java
/**https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/ */
//You are given an m x n integer matrix grid where each cell is either 0 (empty) or 1 (obstacle). You can move up, down, left, or right from and to an empty cell in one step.
//Return the minimum number of steps to walk from the upper left corner (0, 0) to the lower right corner (m - 1, n - 1) given that you can eliminate at most k obstacles. If it is not possible to find such walk return -1.


class ShortestPathInAGridWithObstaclesElimination {
    public int shortestPath(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        int[][][] visited = new int[m][n][k + 1];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                Arrays.fill(visited[i][j], Integer.MAX_VALUE);
            }
        }
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0, 0});
        visited[0][0][0] = 0;
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];
            int step = cur[2];
            if (x == m - 1 && y == n - 1) {
                return step;
            }
            for (int[] dir : new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}) {
                int nx = x + dir[0];
                int ny = y + dir[1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n) {
                    int nstep = step + 1;
                    int nk = grid[nx][ny] == 1 ? k - 1 : k;
                    if (nk >= 0 && nstep < visited[nx][ny][nk]) {
                        visited[nx][ny][nk] = nstep;
                        queue.offer(new int[]{nx, ny, nstep});
                    }
                }
            }
        }
        return -1;
        
    }
}
     
    

File: Tool/PreUtil/src/output/javascript/hard/ShortestPathInAGridWithObstaclesElimination.js
/**https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/ */
//You are given an m x n integer matrix grid where each cell is either 0 (empty) or 1 (obstacle). You can move up, down, left, or right from and to an empty cell in one step.
//Return the minimum number of steps to walk from the upper left corner (0, 0) to the lower right corner (m - 1, n - 1) given that you can eliminate at most k obstacles. If it is not possible to find such walk return -1.


/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number}
 */
var shortestPath = function(grid, k) {
    let n = grid.length;
    let m = grid[0].length;
    let visited = new Set();
    let queue = [[0, 0, 0, 0]];
    while(queue.length > 0){
        let [r, c, moves, obstacles] = queue.shift();
        if(r == n - 1 && c == m - 1){
            return moves;
        }
        let key = getKey(r, c, obstacles);
        if(visited.has(key)){
            continue;
        }
        visited.add(key);
        for(let [nr, nc] of [[r + 1, c], [r - 1, c], [r, c + 1], [r, c - 1]]){
            if(nr >= 0 && nr < n && nc >= 0 && nc < m){
                if(grid[nr][nc] == 0){
                    queue.push([nr, nc, moves + 1, obstacles]);
                }
                else if(obstacles < k){
                    queue.push([nr, nc, moves + 1, obstacles + 1]);
                }
            }
        }
    }
    return -1;
    
};

    